### PR TITLE
*: make rollback work on user-defined variables

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Partly fix https://github.com/pingcap/tidb/issues/8393

Problem Summary:

`set @xx = yy` take effect immediately in TiDB, while in MySQL this statement can be rollbacked.

### What is changed and how it works?

What's Changed:

Make rollback work on user-defined variables.

How it Works:

Use struct `userDefinedVars` for user-defined variables.
This struct consists of a slice and a map, before commit, the modification is cached in the slice.
It has the `Commit` and `Rollback` methods to flush or drop the cached variables.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

### Release note 

- make rollback work on user-defined variables

### CODE REVIEW

**Comment by tiancaiamao** [comment]
PTAL @XuHuaiyu @lzmhhh123

**Comment by tiancaiamao** [comment]
Hi, @morgo 
What's your opinion about this behaviour change?
Should we try to achieve the MySQL compatibility here? this behaviour change may also break the compatibility if we upgrade from an old TiDB cluster ...

File: sessionctx/variable/session.go [hunk 1]
```diff
@@ -360,6 +360,68 @@ const (
...
 )
 
+type userDefinedVars struct {
+	names  []string
+	values []types.Datum
+	// There is no NULL Datum stored, NULL is translated to delete operation.
+	stored map[string]types.Datum
+}
+
+func (udv *userDefinedVars) Get(name string) (types.Datum, bool) {
```
Comment (AilinKid) [inline sessionctx/variable/session.go:370]: will it be concurrent access in the future?

**Review by AilinKid — COMMENTED** [review]
Rest LGTM

**Comment by mjonss** [comment]
User variables are *not* transactional in MySQL:
```sql
mysql> select @testvar;
+--------------------+
| @testvar           |
+--------------------+
| NULL               |
+--------------------+
1 row in set (0,00 sec)

mysql> begin;
Query OK, 0 rows affected (0,00 sec)

mysql> set @testvar = 1;
Query OK, 0 rows affected (0,00 sec)

mysql> rollback;
Query OK, 0 rows affected (0,00 sec)

mysql> select @testvar;
+----------+
| @testvar |
+----------+
|        1 |
+----------+
1 row in set (0,00 sec)

mysql> select version();
+-------------------------+
| version()               |
+-------------------------+
| 8.0.25-0ubuntu0.21.04.1 |
+-------------------------+
1 row in set (0,00 sec)

```

Neither are session or global variables in MySQL.
So I don't think making it transactional in TiDB makes sense, unless it is a requested feature and have its own flag for enable the compatibility breaking feature.

I also added a comment to https://github.com/pingcap/tidb/issues/8393

**Review by mjonss — CHANGES_REQUESTED** [review]
Although the patch seems good, it is based on the assumption that user variables *should* be transactional, when they are in fact *not* transactional in MySQL.

So I don't think we should push this change into TiDB and break the compatibility with MySQL.

**Pull request closed without merging**